[in-progress] turn on user warnings for AO migration

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94071
* #94422
* #94420

Summary:

There is a long standing migration of `torch.quantization` to `torch.ao.quantization`
and `torch.nn.{quantization_related_submodules}` to `torch.ao.nn.{quantization_related_submodules}`.
The eventual goal is to make `torch.ao` self-contained for quantization Python code
in PyTorch. For example, if a user was using `torch.quantization.MinMaxObserver`,
the new location of this is `torch.ao.quantization.MinMaxObserver`, and the old
location still works to maintain BC.

https://github.com/pytorch/pytorch/issues/81667 has context on this. Note: we need
to make that issue more descriptive before turning on the warnings.

Overall status of the effort:
1. [done] migrate PyTorch AO code to new locations, leave old locations as stubs
2. [not done] migrate PyTorch callsites to new locations
3. [this PR, currently RFC] turn on deprecation warnings on old locations
4. [not done, scheduled for PT2.x] delete the old locations

I'd like to get (2) and (3) landed for PT2.0, so we can delete everything in PT2.x.

Note, (2) should be done before (3) to prevent warning spam.  This PR is just
ensuring that we have a plan for (3), it will only be landed after (2)
is also complete.

Now, to the actual goal of this PR:
a. if user accesses an old style name, a warning should be shown
b. if user accesses a new style name, no warning should be shown
c. if user does not use quantization, no warning should be shown

(c) is unfortunately problematic because `torch/__init__.py` imports the
old quantization locations,
so we can't just add warnings to the old locations without spamming everyone
who imports `torch`. Because of this, we use https://peps.python.org/pep-0562/
to patch attribute access of the old locations to print this warning.

Comments welcome - if this approach is good to go, we can apply it to all submodules
in all of the folders being migrated. Keeping it as an example for now so that
review feedback can be addressed faster.

Test plan:

```
>>> import torch

>>> torch.quantization
<module 'torch.quantization' from '/Users/vasiliy/pytorch/torch/quantization/__init__.py'>

>>> torch.quantization.observer
<module 'torch.quantization.observer' from '/Users/vasiliy/pytorch/torch/quantization/observer.py'>

>>> torch.quantization.MinMaxObserver
/Users/vasiliy/pytorch/torch/quantization/__init__.py:78: UserWarning: `torch.quantization.MinMaxObserver` has been moved to `torch.ao.quantization.MinMaxObserver`, and the `torch.quantization.MinMaxObserver` syntax will be removed in a future version of PyTorch. Please see https://github.com/pytorch/pytorch/issues/81667 for detailed migration instructions.
  warnings.warn(_get_ao_migration_warning_str(__name__, name))
  <class 'torch.ao.quantization.observer.MinMaxObserver'>

>>> torch.quantization.observer.MinMaxObserver
/Users/vasiliy/pytorch/torch/quantization/observer.py:54: UserWarning: `torch.quantization.observer.MinMaxObserver` has been moved to `torch.ao.quantization.observer.MinMaxObserver`, and the `torch.quantization.observer.MinMaxObserver` syntax will be removed in a future version of PyTorch. Please see https://github.com/pytorch/pytorch/issues/81667 for detailed migration instructions.
  warnings.warn(_get_ao_migration_warning_str(__name__, name))
  <class 'torch.ao.quantization.observer.MinMaxObserver'>
```

vkuzo committed:
[rfc] turn on user warnings for AO migration

Summary:

There is a long standing migration of `torch.quantization` to `torch.ao.quantization`
and `torch.nn.{quantization_related_submodules}` to `torch.ao.nn.{quantization_related_submodules}`.
The eventual goal is to make `torch.ao` self-contained for quantization Python code
in PyTorch. For example, if a user was using `torch.quantization.MinMaxObserver`,
the new location of this is `torch.ao.quantization.MinMaxObserver`, and the old
location still works to maintain BC.

https://github.com/pytorch/pytorch/issues/81667 has context on this. Note: we need
to make that issue more descriptive before turning on the warnings.

Overall status of the effort:
1. [done] migrate PyTorch AO code to new locations, leave old locations as stubs
2. [not done] migrate PyTorch callsites to new locations
3. [this PR, currently RFC] turn on deprecation warnings on old locations
4. [not done, scheduled for PT2.x] delete the old locations

I'd like to get (2) and (3) landed for PT2.0, so we can delete everything in PT2.x.

Note, (2) should be done before (3) to prevent warning spam.  This PR is just
ensuring that we have a plan for (3), it will only be landed after (2)
is also complete.

Now, to the actual goal of this PR:
a. if user accesses an old style name, a warning should be shown
b. if user accesses a new style name, no warning should be shown
c. if user does not use quantization, no warning should be shown

(c) is unfortunately problematic because `torch/__init__.py` imports the
old quantization locations,
so we can't just add warnings to the old locations without spamming everyone
who imports `torch`. Because of this, we use https://peps.python.org/pep-0562/
to patch attribute access of the old locations to print this warning.

Comments welcome - if this approach is good to go, we can apply it to all submodules
in all of the folders being migrated. Keeping it as an example for now so that
review feedback can be addressed faster.

Test plan:

```
>>> import torch

>>> torch.quantization
<module 'torch.quantization' from '/Users/vasiliy/pytorch/torch/quantization/__init__.py'>

>>> torch.quantization.observer
<module 'torch.quantization.observer' from '/Users/vasiliy/pytorch/torch/quantization/observer.py'>

>>> torch.quantization.MinMaxObserver
/Users/vasiliy/pytorch/torch/quantization/__init__.py:78: UserWarning: `torch.quantization.MinMaxObserver` has been moved to `torch.ao.quantization.MinMaxObserver`, and the `torch.quantization.MinMaxObserver` syntax will be removed in a future version of PyTorch. Please see https://github.com/pytorch/pytorch/issues/81667 for detailed migration instructions.
  warnings.warn(_get_ao_migration_warning_str(__name__, name))
  <class 'torch.ao.quantization.observer.MinMaxObserver'>

>>> torch.quantization.observer.MinMaxObserver
/Users/vasiliy/pytorch/torch/quantization/observer.py:54: UserWarning: `torch.quantization.observer.MinMaxObserver` has been moved to `torch.ao.quantization.observer.MinMaxObserver`, and the `torch.quantization.observer.MinMaxObserver` syntax will be removed in a future version of PyTorch. Please see https://github.com/pytorch/pytorch/issues/81667 for detailed migration instructions.
  warnings.warn(_get_ao_migration_warning_str(__name__, name))
  <class 'torch.ao.quantization.observer.MinMaxObserver'>
```

[ghstack-poisoned]

vkuzo commented on torch/utils/_migration_utils.py:
```diff
@@ -0,0 +1,24 @@
+# backwards compatible, and a low-effort way to do this is to allow
+# the deprecated objects to still participate in the `import *`
+# syntax.
+_AO_MIGRATION_DEPRECATED_NAME_PREFIX = "deprecated"
```
will probably need to add the underscores to this and do all the imports explicitly to make `test_public_bindings` happy

vkuzo committed:
Update on "[rfc] turn on user warnings for AO migration"


Summary:

There is a long standing migration of `torch.quantization` to `torch.ao.quantization`
and `torch.nn.{quantization_related_submodules}` to `torch.ao.nn.{quantization_related_submodules}`.
The eventual goal is to make `torch.ao` self-contained for quantization Python code
in PyTorch. For example, if a user was using `torch.quantization.MinMaxObserver`,
the new location of this is `torch.ao.quantization.MinMaxObserver`, and the old
location still works to maintain BC.

https://github.com/pytorch/pytorch/issues/81667 has context on this. Note: we need
to make that issue more descriptive before turning on the warnings.

Overall status of the effort:
1. [done] migrate PyTorch AO code to new locations, leave old locations as stubs
2. [not done] migrate PyTorch callsites to new locations
3. [this PR, currently RFC] turn on deprecation warnings on old locations
4. [not done, scheduled for PT2.x] delete the old locations

I'd like to get (2) and (3) landed for PT2.0, so we can delete everything in PT2.x.

Note, (2) should be done before (3) to prevent warning spam.  This PR is just
ensuring that we have a plan for (3), it will only be landed after (2)
is also complete.

Now, to the actual goal of this PR:
a. if user accesses an old style name, a warning should be shown
b. if user accesses a new style name, no warning should be shown
c. if user does not use quantization, no warning should be shown

(c) is unfortunately problematic because `torch/__init__.py` imports the
old quantization locations,
so we can't just add warnings to the old locations without spamming everyone
who imports `torch`. Because of this, we use https://peps.python.org/pep-0562/
to patch attribute access of the old locations to print this warning.

Comments welcome - if this approach is good to go, we can apply it to all submodules
in all of the folders being migrated. Keeping it as an example for now so that
review feedback can be addressed faster.

Test plan:

```
>>> import torch

>>> torch.quantization
<module 'torch.quantization' from '/Users/vasiliy/pytorch/torch/quantization/__init__.py'>

>>> torch.quantization.observer
<module 'torch.quantization.observer' from '/Users/vasiliy/pytorch/torch/quantization/observer.py'>

>>> torch.quantization.MinMaxObserver
/Users/vasiliy/pytorch/torch/quantization/__init__.py:78: UserWarning: `torch.quantization.MinMaxObserver` has been moved to `torch.ao.quantization.MinMaxObserver`, and the `torch.quantization.MinMaxObserver` syntax will be removed in a future version of PyTorch. Please see https://github.com/pytorch/pytorch/issues/81667 for detailed migration instructions.
  warnings.warn(_get_ao_migration_warning_str(__name__, name))
  <class 'torch.ao.quantization.observer.MinMaxObserver'>

>>> torch.quantization.observer.MinMaxObserver
/Users/vasiliy/pytorch/torch/quantization/observer.py:54: UserWarning: `torch.quantization.observer.MinMaxObserver` has been moved to `torch.ao.quantization.observer.MinMaxObserver`, and the `torch.quantization.observer.MinMaxObserver` syntax will be removed in a future version of PyTorch. Please see https://github.com/pytorch/pytorch/issues/81667 for detailed migration instructions.
  warnings.warn(_get_ao_migration_warning_str(__name__, name))
  <class 'torch.ao.quantization.observer.MinMaxObserver'>
```

[ghstack-poisoned]

vkuzo committed:
Update on "[rfc] turn on user warnings for AO migration"


Summary:

There is a long standing migration of `torch.quantization` to `torch.ao.quantization`
and `torch.nn.{quantization_related_submodules}` to `torch.ao.nn.{quantization_related_submodules}`.
The eventual goal is to make `torch.ao` self-contained for quantization Python code
in PyTorch. For example, if a user was using `torch.quantization.MinMaxObserver`,
the new location of this is `torch.ao.quantization.MinMaxObserver`, and the old
location still works to maintain BC.

https://github.com/pytorch/pytorch/issues/81667 has context on this. Note: we need
to make that issue more descriptive before turning on the warnings.

Overall status of the effort:
1. [done] migrate PyTorch AO code to new locations, leave old locations as stubs
2. [not done] migrate PyTorch callsites to new locations
3. [this PR, currently RFC] turn on deprecation warnings on old locations
4. [not done, scheduled for PT2.x] delete the old locations

I'd like to get (2) and (3) landed for PT2.0, so we can delete everything in PT2.x.

Note, (2) should be done before (3) to prevent warning spam.  This PR is just
ensuring that we have a plan for (3), it will only be landed after (2)
is also complete.

Now, to the actual goal of this PR:
a. if user accesses an old style name, a warning should be shown
b. if user accesses a new style name, no warning should be shown
c. if user does not use quantization, no warning should be shown

(c) is unfortunately problematic because `torch/__init__.py` imports the
old quantization locations,
so we can't just add warnings to the old locations without spamming everyone
who imports `torch`. Because of this, we use https://peps.python.org/pep-0562/
to patch attribute access of the old locations to print this warning.

Comments welcome - if this approach is good to go, we can apply it to all submodules
in all of the folders being migrated. Keeping it as an example for now so that
review feedback can be addressed faster.

Test plan:

```
>>> import torch

>>> torch.quantization
<module 'torch.quantization' from '/Users/vasiliy/pytorch/torch/quantization/__init__.py'>

>>> torch.quantization.observer
<module 'torch.quantization.observer' from '/Users/vasiliy/pytorch/torch/quantization/observer.py'>

>>> torch.quantization.MinMaxObserver
/Users/vasiliy/pytorch/torch/quantization/__init__.py:78: UserWarning: `torch.quantization.MinMaxObserver` has been moved to `torch.ao.quantization.MinMaxObserver`, and the `torch.quantization.MinMaxObserver` syntax will be removed in a future version of PyTorch. Please see https://github.com/pytorch/pytorch/issues/81667 for detailed migration instructions.
  warnings.warn(_get_ao_migration_warning_str(__name__, name))
  <class 'torch.ao.quantization.observer.MinMaxObserver'>

>>> torch.quantization.observer.MinMaxObserver
/Users/vasiliy/pytorch/torch/quantization/observer.py:54: UserWarning: `torch.quantization.observer.MinMaxObserver` has been moved to `torch.ao.quantization.observer.MinMaxObserver`, and the `torch.quantization.observer.MinMaxObserver` syntax will be removed in a future version of PyTorch. Please see https://github.com/pytorch/pytorch/issues/81667 for detailed migration instructions.
  warnings.warn(_get_ao_migration_warning_str(__name__, name))
  <class 'torch.ao.quantization.observer.MinMaxObserver'>
```

[ghstack-poisoned]

vkuzo commented:
ok, CI is green (with (2) not done yet, so we're getting some testing for free), and I haven't heard concerns, assuming we can use this approach.

albanD reviewed (commented):
Approach sounds good!

vkuzo committed:
Update on "[rfc] turn on user warnings for AO migration"


Summary:

There is a long standing migration of `torch.quantization` to `torch.ao.quantization`
and `torch.nn.{quantization_related_submodules}` to `torch.ao.nn.{quantization_related_submodules}`.
The eventual goal is to make `torch.ao` self-contained for quantization Python code
in PyTorch. For example, if a user was using `torch.quantization.MinMaxObserver`,
the new location of this is `torch.ao.quantization.MinMaxObserver`, and the old
location still works to maintain BC.

https://github.com/pytorch/pytorch/issues/81667 has context on this. Note: we need
to make that issue more descriptive before turning on the warnings.

Overall status of the effort:
1. [done] migrate PyTorch AO code to new locations, leave old locations as stubs
2. [not done] migrate PyTorch callsites to new locations
3. [this PR, currently RFC] turn on deprecation warnings on old locations
4. [not done, scheduled for PT2.x] delete the old locations

I'd like to get (2) and (3) landed for PT2.0, so we can delete everything in PT2.x.

Note, (2) should be done before (3) to prevent warning spam.  This PR is just
ensuring that we have a plan for (3), it will only be landed after (2)
is also complete.

Now, to the actual goal of this PR:
a. if user accesses an old style name, a warning should be shown
b. if user accesses a new style name, no warning should be shown
c. if user does not use quantization, no warning should be shown

(c) is unfortunately problematic because `torch/__init__.py` imports the
old quantization locations,
so we can't just add warnings to the old locations without spamming everyone
who imports `torch`. Because of this, we use https://peps.python.org/pep-0562/
to patch attribute access of the old locations to print this warning.

Comments welcome - if this approach is good to go, we can apply it to all submodules
in all of the folders being migrated. Keeping it as an example for now so that
review feedback can be addressed faster.

Test plan:

```
>>> import torch

>>> torch.quantization
<module 'torch.quantization' from '/Users/vasiliy/pytorch/torch/quantization/__init__.py'>

>>> torch.quantization.observer
<module 'torch.quantization.observer' from '/Users/vasiliy/pytorch/torch/quantization/observer.py'>

>>> torch.quantization.MinMaxObserver
/Users/vasiliy/pytorch/torch/quantization/__init__.py:78: UserWarning: `torch.quantization.MinMaxObserver` has been moved to `torch.ao.quantization.MinMaxObserver`, and the `torch.quantization.MinMaxObserver` syntax will be removed in a future version of PyTorch. Please see https://github.com/pytorch/pytorch/issues/81667 for detailed migration instructions.
  warnings.warn(_get_ao_migration_warning_str(__name__, name))
  <class 'torch.ao.quantization.observer.MinMaxObserver'>

>>> torch.quantization.observer.MinMaxObserver
/Users/vasiliy/pytorch/torch/quantization/observer.py:54: UserWarning: `torch.quantization.observer.MinMaxObserver` has been moved to `torch.ao.quantization.observer.MinMaxObserver`, and the `torch.quantization.observer.MinMaxObserver` syntax will be removed in a future version of PyTorch. Please see https://github.com/pytorch/pytorch/issues/81667 for detailed migration instructions.
  warnings.warn(_get_ao_migration_warning_str(__name__, name))
  <class 'torch.ao.quantization.observer.MinMaxObserver'>
```

[ghstack-poisoned]

vkuzo committed:
Update on "[rfc] turn on user warnings for AO migration"


Summary:

There is a long standing migration of `torch.quantization` to `torch.ao.quantization`
and `torch.nn.{quantization_related_submodules}` to `torch.ao.nn.{quantization_related_submodules}`.
The eventual goal is to make `torch.ao` self-contained for quantization Python code
in PyTorch. For example, if a user was using `torch.quantization.MinMaxObserver`,
the new location of this is `torch.ao.quantization.MinMaxObserver`, and the old
location still works to maintain BC.

https://github.com/pytorch/pytorch/issues/81667 has context on this. Note: we need
to make that issue more descriptive before turning on the warnings.

Overall status of the effort:
1. [done] migrate PyTorch AO code to new locations, leave old locations as stubs
2. [not done] migrate PyTorch callsites to new locations
3. [this PR, currently RFC] turn on deprecation warnings on old locations
4. [not done, scheduled for PT2.x] delete the old locations

I'd like to get (2) and (3) landed for PT2.0, so we can delete everything in PT2.x.

Note, (2) should be done before (3) to prevent warning spam.  This PR is just
ensuring that we have a plan for (3), it will only be landed after (2)
is also complete.

Now, to the actual goal of this PR:
a. if user accesses an old style name, a warning should be shown
b. if user accesses a new style name, no warning should be shown
c. if user does not use quantization, no warning should be shown

(c) is unfortunately problematic because `torch/__init__.py` imports the
old quantization locations,
so we can't just add warnings to the old locations without spamming everyone
who imports `torch`. Because of this, we use https://peps.python.org/pep-0562/
to patch attribute access of the old locations to print this warning.

Comments welcome - if this approach is good to go, we can apply it to all submodules
in all of the folders being migrated. Keeping it as an example for now so that
review feedback can be addressed faster.

Test plan:

```
>>> import torch

>>> torch.quantization
<module 'torch.quantization' from '/Users/vasiliy/pytorch/torch/quantization/__init__.py'>

>>> torch.quantization.observer
<module 'torch.quantization.observer' from '/Users/vasiliy/pytorch/torch/quantization/observer.py'>

>>> torch.quantization.MinMaxObserver
/Users/vasiliy/pytorch/torch/quantization/__init__.py:78: UserWarning: `torch.quantization.MinMaxObserver` has been moved to `torch.ao.quantization.MinMaxObserver`, and the `torch.quantization.MinMaxObserver` syntax will be removed in a future version of PyTorch. Please see https://github.com/pytorch/pytorch/issues/81667 for detailed migration instructions.
  warnings.warn(_get_ao_migration_warning_str(__name__, name))
  <class 'torch.ao.quantization.observer.MinMaxObserver'>

>>> torch.quantization.observer.MinMaxObserver
/Users/vasiliy/pytorch/torch/quantization/observer.py:54: UserWarning: `torch.quantization.observer.MinMaxObserver` has been moved to `torch.ao.quantization.observer.MinMaxObserver`, and the `torch.quantization.observer.MinMaxObserver` syntax will be removed in a future version of PyTorch. Please see https://github.com/pytorch/pytorch/issues/81667 for detailed migration instructions.
  warnings.warn(_get_ao_migration_warning_str(__name__, name))
  <class 'torch.ao.quantization.observer.MinMaxObserver'>
```

[ghstack-poisoned]

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -6,31 +6,51 @@
+    setattr(sys.modules[__name__], target_obj_name, target_obj)
+
+def __getattr__(name):
+    if name in _deprecated_names:
```
Python does it for us for free, which is nice
